### ISSUE TITLE
[CI] failing tests RBProgramNodeTest

### ISSUE THREAD

**Issue by MarcusDenker** [guillep/pharo-core#16160]
**Bug description**

three tests are failing:

- testIsA – MacOSX64.AST.Core.Tests.RBParserTest
- testIsUsedAsReturnValueWhenNodeIsLastInBlockExpectTrue – MacOSX64.AST.Core.Tests.RBProgramNodeTest
- testlastIsReturnWhenArrayNodeWithLastReturnExpectFalse – MacOSX64.AST.Core.Tests.RBProgramNodeTest

I think this is due to merging https://github.com/pharo-project/pharo/pull/16146

**Comment by balsa-sarenac** [comment]
Thanks Marcus, I'll see to fix remaining failing tests today or tomorrow.

**Comment by balsa-sarenac** [comment]
@Ducasse @MarcusDenker I'm working on this now and I would like to consult you on `RBParserTest`:

The failing test checks how many nodes are essential (previously `isUsed`). The tree in question is:
```
method: arg1
 | temps |
 temps := #(10).
 temps
   foo;
   foo.
 ^(temps collect: [:e | ])
```

old `isUsed` implementation returned:
```
RBVariableNode(arg1)
RBSequenceNode(| temps |
temps := #( 10 ).
temps
	foo;
	foo.
^ temps collect: [ :e |  ])
RBLiteralArrayNode(#( 10 ))
RBLiteralValueNode(10)
RBVariableNode(temps)
RBVariableNode(temps)
RBMessageNode(temps collect: [ :e |  ])
RBVariableNode(temps)
RBBlockNode([ :e |  ])
```

new implementation with `isEssential` returns:
```
RBVariableNode(arg1)
RBSequenceNode(| temps |
temps := #( 10 ).
temps
	foo;
	foo.
^ temps collect: [ :e |  ])
RBLiteralArrayNode(#( 10 ))
RBLiteralValueNode(10)
RBVariableNode(temps)
RBVariableNode(temps)
RBVariableNode(temps)
RBMessageNode(temps collect: [ :e |  ])
RBVariableNode(temps)
RBBlockNode([ :e |  ])
RBVariableNode(e)
RBSequenceNode()
```

You can compare implementors of `uses:` from an older version of Pharo with new implementors of `isEssentialChild:` to find the implementation differences.
Basically the new implementation says that everything that is not child of sequence is not essential and can not be removed.

I've fixed the test so it passes. Based on your answers we can modify the implementation and the tests later as well.

**Comment by Ducasse** [comment]
To me I would say that Receiver Msgs: Arg, x := Value, ^ X and [ LAST ] are essential 
So all the rest is not essential: statements of sequence except the last one in a block. no?

**Comment by balsa-sarenac** [comment]
That's like we talked, except that we said last sequence in block is not essential . This is what I get for not essential nodes:

Old implementation:
```
method: arg1

	| temps |
	temps := #( 10 ).
	temps
		foo;
		foo.
	^ temps collect: [ :e |  ]
RBVariableNode(arg1)
RBVariableNode(temps)
RBAssignmentNode(temps := #( 10 ))
RBVariableNode(temps)
RBCascadeNode(temps
	foo;
	foo)
RBMessageNode(temps foo)
RBMessageNode(temps foo)
RBVariableNode(e)
```

New implementation:
```
method: arg1

	| temps |
	temps := #( 10 ).
	temps
		foo;
		foo.
	^ temps collect: [ :e |  ]
RBVariableNode(temps)
RBAssignmentNode(temps := #( 10 ))
RBCascadeNode(temps
	foo;
	foo)
RBMessageNode(temps foo)
RBMessageNode(temps foo)
RBReturnNode(^ temps collect: [ :e |  ])
```
So the new one does exactly that: all statements except for the last one in the block.
